umbra-js/src/utils/cns.ts

```diff
@@ -1,57 +1,100 @@
+import type { EthersProvider, TransactionResponse } from '../types';
 import * as cnsResolverAbi from '../abi/CnsResolver.json';
 import { createContract } from './utils';
 
-const { CNS_REGISTRY } = constants;
+export const cnsKeyPathSpending = 'crypto.ETH.umbra.spending_public_key';
+export const cnsKeyPathViewing = 'crypto.ETH.umbra.viewing_public_key';
+
+/**
+ * @notice Returns supported CNS domain endings
+ */
+export const supportedCnsDomains = ['.crypto', '.zil'];
 
-export const umbraKeySignature = 'crypto.ETH.signature-v0';
+/**
+ * @notice Returns true if the provided name is an CNS domain, false otherwise
+ * @param domainName Name to check
+ */
+export function isCnsDomain(domainName: string) {
+  for (const suffix of supportedCnsDomains) {
+    if (domainName.endsWith(suffix)) {
+      return true;
+    }
+  }
+  return false;
+}
 
 /**
  * @notice Computes CNS namehash of the input CNS domain, normalized to CNS compatibility
  * @param name CNS domain, e.g. myname.crypto
  * @param resolution Resolution instance of @unstoppabledomains/resolution
  */
 export function namehash(name: string, resolution: Resolution) {
+  if (typeof name !== 'string') {
+    throw new Error('Name must be a string with a supported suffix');
+  }
+  if (!isCnsDomain(name)) {
+    throw new Error(`Name does not end with supported suffix: ${supportedCnsDomains.join(', ')}`);
+  }
   return resolution.namehash(name);
 }
 
 /**
- * @notice For a given CNS domain, recovers and returns the public key from its signature
+ * @notice For a given ENS domain, returns the public keys, or throws if they don't exist
  * @param name CNS domain, e.g. myname.crypto
+ * @param provider ethers provider to use
  * @param resolution Resolution instance of @unstoppabledomains/resolution
  */
-export function getPublicKeys(name: string, resolution: Resolution) {
-  name; // silence errors
-  resolution; // silence errors
-  const spendingPublicKey = '';
-  const viewingPublicKey = '';
-
+export async function getPublicKeys(name: string, provider: EthersProvider, resolution: Resolution) {
+  // Read compressed public keys
+  const domainNamehash = resolution.namehash(name);
+  const resolverAddress = await resolution.resolver(name);
+  const cnsResolver = createContract(resolverAddress, cnsResolverAbi, provider);
+  const [compressedSpendingPublicKey, compressedViewingPublicKey] = await cnsResolver.getMany(
+    [cnsKeyPathSpending, cnsKeyPathViewing],
+    domainNamehash
+  );
+  if (!compressedSpendingPublicKey || !compressedViewingPublicKey) {
+    throw new Error(`Public keys not found for ${name}. User must setup their Umbra account`);
+  }
+  // Return uncompressed public keys
+  const spendingPublicKey = computePublicKey(compressedSpendingPublicKey);
+  const viewingPublicKey = computePublicKey(compressedViewingPublicKey);
   return { spendingPublicKey, viewingPublicKey };
 }
 
 /**
  * @notice For a given CNS domain, sets the associated umbra signature
```

Comment:
This say "sets the associated umbra _signature_" but I believe that's outdated, as we're just setting the two public keys directly now.

Reply:
Fixed in 1723291